components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/user/impl/UserInfoJWTResponse.java

```diff
@@ -67,7 +79,49 @@ public String getResponseString(OAuth2TokenValidationResponseDTO tokenResponse)
 
         JWTClaimsSet jwtClaimsSet = new JWTClaimsSet();
         jwtClaimsSet.setAllClaims(claims);
-        return new PlainJWT(jwtClaimsSet).serialize();
+
+        String sigAlg = OAuthServerConfiguration.getInstance().getUserInfoJWTSignatureAlgorithm();
+        if (sigAlg != null && !sigAlg.trim().isEmpty()) {
+            try {
+                signatureAlgorithm = OAuth2Util.mapSignatureAlgorithm(sigAlg);
+            } catch (IdentityOAuth2Exception e) {
+                throw new UserInfoEndpointException("Unsupported signature algorithm configured.", e);
+            }
+        }
+
+        if (JWSAlgorithm.NONE.getName().equals(signatureAlgorithm.getName())) {
+            return new PlainJWT(jwtClaimsSet).serialize();
+        }
+
+        boolean isJWTSignedWithSPKey = OAuthServerConfiguration.getInstance().isJWTSignedWithSPKey();
+        String tenantDomain;
+        if (isJWTSignedWithSPKey) {
+            try {
+                AccessTokenDO accessTokenDO = OAuth2Util.getAccessTokenDOfromTokenIdentifier(tokenResponse
+                        .getAuthorizationContextToken().getTokenString());
+                String clientId;
+                if (accessTokenDO != null) {
+                    clientId = accessTokenDO.getConsumerKey();
+                } else {
+                    // this means the token is not active so we can't proceed further
+                    throw new UserInfoEndpointException(OAuthError.ResourceResponse.INVALID_TOKEN,
+                            "Invalid Access Token. Access token is not ACTIVE.");
+                }
+
+                OAuthAppDO oAuthAppDO = OAuth2Util.getAppInformationByClientId(clientId);
+                tenantDomain = OAuth2Util.getTenantDomainOfOauthApp(oAuthAppDO);
+            } catch (IdentityOAuth2Exception | InvalidOAuthClientException e) {
+                throw new UserInfoEndpointException("Error occurred while signing JWT", e);
+            }
+        } else {
+            tenantDomain = MultitenantUtils.getTenantDomain(tokenResponse.getAuthorizedUser());
+        }
+
+        try {
+            return OAuth2Util.signJWT(jwtClaimsSet, signatureAlgorithm, tenantDomain);
```

Comment:
There's only one property can be configured in identity.xml as follows to specify the userinfo JWT response signing algorithm. If that property is not available/property value is null, by default it will not sign the response (send as plain JWT). This is already implemented above.
<OAuth>
    …….
     <OpenIDConnect> 
          <UserInfoJWTSignatureAlgorithm>false</UserInfoJWTSignatureAlgorithm>
         …...
     </OpenIDConnect> 
</OAuth>